src/fortran_routing/diffusive_v02frwk/diffusive_cnt/diffusive_cnt.f90

```diff
@@ -251,72 +240,62 @@ subroutine diffnw(dtini_g, t0_g, tfin_g, saveinterval_ev_g, dt_ql_g, dt_ub_g, dt
             end if
-            
-            !* downstream boundary condition for TW reach
             if (frnw_g(j,2)<0.0) then
-                !* 1. measured data
+            !* downstream boundary water elevation at bottom node of TW reach                
+		!* 1. measured data
 !                do n=1,nts_db_g
 !                    varr_db(n)= dbcd_g(n) + z(ncomp,j) !* when dbcd is water stage, channel bottom elev is added.
 !                enddo
 !                oldY(ncomp,j)= intp_y(nts_db_g, tarr_db, varr_db, t)
                 ! 2. normal depth as TW boundary condition
-                q_sk_multi=1.0
                 slope = (z(ncomp-1,j)-z(ncomp,j))/dx(ncomp-1,j)
-                if (slope .le. 0.0001) slope = 0.0001
-		oldQ(ncomp,j)= iniq(ncomp,j)
-                !oldY below takes normal depth value as a result.
+                if (slope .le. 0.0001) slope = 0.0001		
+                !* oldY(i,j) <- normal depth for iniq(i,j) + z(i,j). Hence, oldY is water elevation [m].
                 call normal_crit_y(ncomp, j, q_sk_multi, slope, oldQ(ncomp,j), oldY(ncomp,j), temp,  oldArea(ncomp,j), temp)
-            end if
-        end do
-        ! +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
-        ! compute initial water depth, celerity, and diffusivity using iniq
-        ! +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
-        q_sk_multi=1.0
-        do j=1, nlinks
-            ncomp= frnw_g(j,1)
-            do i=1, ncomp
-                oldQ(i,j) = iniq(i,j)
-		newQ(i,1,j) = iniq(i,j)
-                qp(i,1,j)= iniq(i,j)
-                if (i==ncomp) then
-                    !* normal depth
-                    slope = (z(i-1,j)-z(i,j))/dx(i-1,j)
-                    if (slope .le. 0.0001) slope = 0.0001
-                    if ((frnw_g(j,2)<0.0).and.(i==ncomp)) then
-                        !*use TW boundary water elevation data
-                        oldY(ncomp,j)=oldY(ncomp,j)
-                    else
-                        !* oldY(i,j) <- normal depth for iniq(i,j) + z(i,j). Hence, oldY is water elevation [m].
-                        call normal_crit_y(i, j, q_sk_multi, slope, oldQ(i,j), oldY(i,j), temp, temp, temp)
-                    endif
-                endif
-            enddo
-            !* for mainly computing celerity and diffusivity
+            else            
+		!* Not TW reach -> water elev at bottom node of u/s reach is equal to that at top node of d/s reach
+		linknb=frnw_g(j,2) ! reach downstream of j
+		oldY(ncomp,j)= newY(1,linknb)   ! taking the WL from the d/s reach
+            endif            
+            !* compute water elev. from top to the second last node and celerity and diffusivity for all the nodes of reach j
 	    newY(ncomp,j)= oldY(ncomp,j)
 	    idmy1=0
             call mesh_diffusive_backward(j,ntim, idmy1)
-            do i=1,ncomp
+            do i=1,ncomp-1
```

Comment:
Up to here, get two for loops, one for boundary conditions at headwater and TW reaches and the other for computing initial Y,C,D, into one for loop to firm it up.